File: app/src/main/java/com/elementary/tasks/core/contacts/CallsFragment.java
package com.elementary.tasks.core.contacts;

import android.app.Activity;
import android.app.Fragment;
import android.content.Context;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.InputMethodManager;

import com.elementary.tasks.core.file_explorer.FilterCallback;
import com.elementary.tasks.core.file_explorer.RecyclerClickListener;
import com.elementary.tasks.databinding.FragmentCallsBinding;

import java.util.List;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class CallsFragment extends Fragment implements CallsLogListener {

    private Context mContext;
    private NumberCallback mCallback;

    private List<CallsItem> mDataList;
    private CallsRecyclerAdapter mAdapter;

    private FragmentCallsBinding binding;
    private RecyclerView mRecyclerView;

    private RecyclerClickListener mClickListener = new RecyclerClickListener() {
        @Override
        public void onItemClick(int position) {
            if (position != -1) {
                String number = mAdapter.getItem(position).getNumber();
                String name = mAdapter.getItem(position).getName();
                if (mCallback != null) {
                    mCallback.onContactSelected(number, name);
                }
            }
        }
    };
    private FilterCallback mFilterCallback = new FilterCallback() {
        @Override
        public void filter(int size) {
            mRecyclerView.scrollToPosition(0);
            refreshView(size);
        }
    };

    public CallsFragment() {
    }

    public static CallsFragment newInstance() {
        return new CallsFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (mContext == null) {
            mContext = context;
        }
        if (mCallback == null) {
            try {
                mCallback = (NumberCallback) context;
            } catch (ClassCastException e) {
                throw new ClassCastException();
            }
        }
    }

    @Override
    public void onAttach(Activity activity) {
        super.onAttach(activity);
        if (mContext == null) {
            mContext = activity;
        }
        if (mCallback == null) {
            try {
                mCallback = (NumberCallback) activity;
            } catch (ClassCastException e) {
                throw new ClassCastException();
            }
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = FragmentCallsBinding.inflate(inflater, container, false);
        initSearchView();
        initRecyclerView();
        binding.refreshLayout.setRefreshing(true);
        binding.refreshLayout.setOnRefreshListener(this::loadData);
        loadData();
        return binding.getRoot();
    }

    private void loadData() {
        new CallsAsync(mContext, this).execute();
    }

    private void initRecyclerView() {
        mRecyclerView = binding.contactsList;
        mRecyclerView.setLayoutManager(new LinearLayoutManager(mContext));
        mRecyclerView.setHasFixedSize(true);
    }

    private void initSearchView() {
        binding.searchField.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (mAdapter != null) mAdapter.filter(s.toString(), mDataList);
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });
    }

    private void refreshView(int count) {
        if (count > 0) {
            binding.emptyItem.setVisibility(View.GONE);
            mRecyclerView.setVisibility(View.VISIBLE);
        } else {
            binding.emptyItem.setVisibility(View.VISIBLE);
            mRecyclerView.setVisibility(View.GONE);
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        InputMethodManager imm = (InputMethodManager) mContext.getSystemService(Context.INPUT_METHOD_SERVICE);
        imm.hideSoftInputFromWindow(binding.searchField.getWindowToken(), 0);
    }

    @Override
    public void onLoaded(List<CallsItem> list) {
        this.mDataList = list;
        binding.refreshLayout.setRefreshing(false);
        mAdapter = new CallsRecyclerAdapter(mContext, list, mClickListener, mFilterCallback);
        mRecyclerView.setAdapter(mAdapter);
        refreshView(mAdapter.getItemCount());
    }
}

File: app/src/main/java/com/elementary/tasks/core/contacts/CallsFragment.kt
package com.elementary.tasks.core.contacts

import android.app.Activity
import android.content.Context
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.inputmethod.InputMethodManager

import com.elementary.tasks.core.file_explorer.FilterCallback
import com.elementary.tasks.core.file_explorer.RecyclerClickListener
import com.elementary.tasks.databinding.FragmentCallsBinding

import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

/**
 * Copyright 2016 Nazar Suhovich
 *
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
class CallsFragment : Fragment(), CallsLogListener {

    private var mContext: Context? = null
    private var mCallback: NumberCallback? = null

    private var mDataList: List<CallsItem>? = null
    private var mAdapter: CallsRecyclerAdapter? = null

    private var binding: FragmentCallsBinding? = null
    private var mRecyclerView: RecyclerView? = null

    private val mClickListener = RecyclerClickListener { position ->
        if (position != -1) {
            val number = mAdapter!!.getItem(position)!!.number
            val name = mAdapter!!.getItem(position)!!.name
            if (mCallback != null) {
                mCallback!!.onContactSelected(number, name)
            }
        }
    }
    private val mFilterCallback = FilterCallback { size ->
        mRecyclerView!!.scrollToPosition(0)
        refreshView(size)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (mContext == null) {
            mContext = context
        }
        if (mCallback == null) {
            try {
                mCallback = context as NumberCallback?
            } catch (e: ClassCastException) {
                throw ClassCastException()
            }

        }
    }

    override fun onAttach(activity: Activity?) {
        super.onAttach(activity)
        if (mContext == null) {
            mContext = activity
        }
        if (mCallback == null) {
            try {
                mCallback = activity as NumberCallback?
            } catch (e: ClassCastException) {
                throw ClassCastException()
            }

        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentCallsBinding.inflate(inflater, container, false)
        initSearchView()
        initRecyclerView()
        binding!!.refreshLayout.isRefreshing = true
        binding!!.refreshLayout.setOnRefreshListener(OnRefreshListener { this.loadData() })
        loadData()
        return binding!!.root
    }

    private fun loadData() {
        CallsAsync(mContext, this).execute()
    }

    private fun initRecyclerView() {
        mRecyclerView = binding!!.contactsList
        mRecyclerView!!.layoutManager = LinearLayoutManager(mContext)
        mRecyclerView!!.setHasFixedSize(true)
    }

    private fun initSearchView() {
        binding!!.searchField.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {

            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (mAdapter != null) mAdapter!!.filter(s.toString(), mDataList)
            }

            override fun afterTextChanged(s: Editable) {

            }
        })
    }

    private fun refreshView(count: Int) {
        if (count > 0) {
            binding!!.emptyItem.visibility = View.GONE
            mRecyclerView!!.visibility = View.VISIBLE
        } else {
            binding!!.emptyItem.visibility = View.VISIBLE
            mRecyclerView!!.visibility = View.GONE
        }
    }

    override fun onPause() {
        super.onPause()
        val imm = mContext!!.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        imm?.hideSoftInputFromWindow(binding!!.searchField.windowToken, 0)
    }

    override fun onLoaded(list: List<CallsItem>) {
        this.mDataList = list
        binding!!.refreshLayout.isRefreshing = false
        mAdapter = CallsRecyclerAdapter(mContext, list, mClickListener, mFilterCallback)
        mRecyclerView!!.adapter = mAdapter
        refreshView(mAdapter!!.itemCount)
    }

    companion object {

        fun newInstance(): CallsFragment {
            return CallsFragment()
        }
    }
}
